tests/integration/callbacks/test_multiple_callbacks.py

```diff
@@ -305,3 +305,281 @@ def out(v):
     # as of https://github.com/plotly/dash/issues/1223, above-out would be 0
     dash_duo.wait_for_text_to_equal("#above-out", "42")
     dash_duo.wait_for_text_to_equal("#below-out", "44")
+
+
+def test_cbmt008_direct_chain(dash_duo):
+    app = dash.Dash(__name__)
+    app.layout = html.Div(
+        [
+            dcc.Input(id="input-1", value="input 1"),
+            dcc.Input(id="input-2"),
+            html.Div("test", id="output"),
+        ]
+    )
+
+    call_counts = {"output": Value("i", 0), "input-2": Value("i", 0)}
+
+    @app.callback(Output("input-2", "value"), Input("input-1", "value"))
+    def update_input(input1):
+        call_counts["input-2"].value += 1
+        return "<<{}>>".format(input1)
+
+    @app.callback(
+        Output("output", "children"),
+        Input("input-1", "value"),
+        Input("input-2", "value"),
+    )
+    def update_output(input1, input2):
+        call_counts["output"].value += 1
+        return "{} + {}".format(input1, input2)
+
+    dash_duo.start_server(app)
+
+    dash_duo.wait_for_text_to_equal("#input-1", "input 1")
+    dash_duo.wait_for_text_to_equal("#input-2", "<<input 1>>")
+    dash_duo.wait_for_text_to_equal("#output", "input 1 + <<input 1>>")
+    assert call_counts["output"].value == 1
+    assert call_counts["input-2"].value == 1
+
+    dash_duo.find_element("#input-1").send_keys("x")
+    dash_duo.wait_for_text_to_equal("#input-1", "input 1x")
+    dash_duo.wait_for_text_to_equal("#input-2", "<<input 1x>>")
+    dash_duo.wait_for_text_to_equal("#output", "input 1x + <<input 1x>>")
+    assert call_counts["output"].value == 2
+    assert call_counts["input-2"].value == 2
+
+    dash_duo.find_element("#input-2").send_keys("y")
+    dash_duo.wait_for_text_to_equal("#input-2", "<<input 1x>>y")
+    dash_duo.wait_for_text_to_equal("#output", "input 1x + <<input 1x>>y")
+    dash_duo.wait_for_text_to_equal("#input-1", "input 1x")
+    assert call_counts["output"].value == 3
+    assert call_counts["input-2"].value == 2
+
+
+def test_cbmt009_branched_chain(dash_duo):
+    app = dash.Dash(__name__)
+    app.layout = html.Div(
+        [
+            dcc.Input(id="grandparent", value="input 1"),
+            dcc.Input(id="parent-a"),
+            dcc.Input(id="parent-b"),
+            html.Div(id="child-a"),
+            html.Div(id="child-b"),
+        ]
+    )
+    call_counts = {
+        "parent-a": Value("i", 0),
+        "parent-b": Value("i", 0),
+        "child-a": Value("i", 0),
+        "child-b": Value("i", 0),
+    }
+
+    @app.callback(Output("parent-a", "value"), Input("grandparent", "value"))
+    def update_parenta(value):
+        call_counts["parent-a"].value += 1
+        return "a: {}".format(value)
+
+    @app.callback(Output("parent-b", "value"), Input("grandparent", "value"))
+    def update_parentb(value):
+        time.sleep(0.2)
```

Comment:
Why the sleep? Is something in this test time senstive?

Reply:
This was just copied over from `test_chained_dependencies_branched_lineage` in the now-deleted [`test_render.py`](https://github.com/plotly/dash/pull/1506/files#diff-cd7e55c505e7622e615c748b9d7ebd9f71d1360e357d9e09b5db946d5cb5a8c0L252) - some of the other tests that had delays embedded I converted to use `Lock`. This one wasn't so obvious to me how to do that, and honestly I didn't look into exactly what it's using the delay for, but I did cut the delays in half.

Reply:
👌